lib/fluent/plugin/bigquery/bigquery_client.rb

```diff
@@ -0,0 +1,138 @@
+require 'json'
+require 'google/api_client'
+require 'google/api_client/client_secrets'
+require 'google/api_client/auth/installed_app'
+require 'google/api_client/auth/compute_service_account'
+
+module Fluent::BigQueryPlugin
+  class BigQueryClient
+    def initialize(attributes = {})
+      attributes.each { |name, value| instance_variable_set("@#{name}", value) }
+    end
+
+    def create_table(name, schema)
+      result =
+        access_api(
+          api_method: bigquery.tables.insert,
+          body_object: {
+            'tableReference' => {
+               'tableId' => name
+            },
+            'schema' => {
+              'fields' => schema
+            }
+          }
+        )
+      handle_error(result) if result.error?
+    end
+
+    def insert(table, rows)
+      result =
+        access_api(
+          api_method: bigquery.tabledata.insert_all,
+          parameters: {
+            'tableId' => table
+          },
+          body_object: {
+            'rows' => rows
+          }
+        )
+      handle_error(result) if result.error?
+    end
+
+    def fetch_schema(table)
+      result =
+        access_api(
+          api_method: bigquery.tables.get,
+          parameters: {
+            'tableId' => table
+          }
+        )
+      handle_error(result) if result.error?
+      JSON.parse(result.body)['schema']['fields']
+    end
+
+    def load
+      # https://developers.google.com/bigquery/loading-data-into-bigquery#loaddatapostrequest
+      raise NotImplementedError # TODO
+    end
+
+    private
+
+    def access_api(params = {})
+      params[:parameters] ||= {}
+      params[:parameters]['projectId'] ||= @project
+      params[:parameters]['datasetId'] ||= @dataset
+      client.execute(params)
+    end
+
+    def bigquery
+      # TODO: refresh with specified expiration
+      @bigquery ||= client.discovered_api('bigquery', 'v2')
+    end
+
+    def handle_error(result)
+      @client = nil # clear cashed client when errors occur
+      error =
+        case result.status
+        when 404      then NotFound
+        when 409      then Conflict
+        when 400..499 then ClientError
+        when 500..599 then ServerError
+        else UnexpectedError
+        end
+      fail error, result.error_message
```

Comment:
google-api-ruby-client has improved, and we can get error message simply.
ref. https://github.com/google/google-api-ruby-client/commit/1ff1a8e355132fdbb4ab1d7da215fb68216418be